controllers/serviceaccount/serviceaccount_controller.go

```diff
@@ -0,0 +1,175 @@
+type ServiceAccountReconciler struct {
+	client.Client
+	ClientSet    *kubernetes.Clientset
+	Scheme       *runtime.Scheme
+	MCClient     client.Client
+	HCPNamespace string
+	log          logr.Logger
+}
+
+func (r *ServiceAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	r.log = ctrllog.FromContext(ctx).WithName("hostedcluster-service-account-controller")
+	r.log.V(1).Info("start reconcile", "Name", req.NamespacedName)
+	serviceAccount := &corev1.ServiceAccount{}
+
+	if r.HCPNamespace == "" {
+		return ctrl.Result{}, nil
+	}
+
+	serviceAccountExists := false
+	// Getting the hosted cluster service account
+	apiContext, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+	err := r.Get(
+		apiContext,
+		types.NamespacedName{Name: constants.MintServiceAccountName, Namespace: constants.HLFWatchedNamespace},
+		serviceAccount,
+	)
+
+	if err != nil && errors.IsNotFound(err) {
+		serviceAccountExists = false
+	} else if err == nil {
+		serviceAccountExists = true
+	} else {
+		return ctrl.Result{RequeueAfter: constants.TokenRefreshDuration}, nil
+	}
+
+	if serviceAccountExists {
+		r.log.V(1).Info("Found existing service account", "Name", serviceAccount.GetName())
+	} else {
+		serviceAccount := &corev1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      constants.MintServiceAccountName,
+				Namespace: constants.HLFWatchedNamespace,
+			},
+		}
+
+		// Create the service account
+		apiContext, cancel := context.WithTimeout(ctx, 10*time.Second)
+		defer cancel()
+
+		if err := r.Create(apiContext, serviceAccount); err != nil {
+			return ctrl.Result{RequeueAfter: time.Minute * 10}, nil
+		}
+	}
+
+	// create service account token
+	token, err := r.createMintServiceAccountToken(ctx, serviceAccount)
+	r.log.V(1).Info("Refresh token", "token", token)
+
+	//copy token to secret
+	r.updateSecretToken(ctx, token)
+
+	//recreate token after
+	return ctrl.Result{RequeueAfter: constants.TokenRefreshDuration}, nil
+}
+
+func (r *ServiceAccountReconciler) createMintServiceAccountToken(
+	ctx context.Context,
+	serviceAccount *corev1.ServiceAccount,
+) (refreshToken string, err error) {
+
+	treq := &authenticationv1.TokenRequest{
+		Spec: authenticationv1.TokenRequestSpec{
+			Audiences: []string{"openshift"},
+		},
+	}
+
+	// Create the service account token
+	apiContext, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	token, err := r.ClientSet.CoreV1().ServiceAccounts(serviceAccount.GetNamespace()).CreateToken(apiContext, serviceAccount.GetName(), treq, metav1.CreateOptions{})
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			r.log.V(1).Info("Token already exists", "Name", token.GetName())
+		} else {
+			return "", fmt.Errorf("failed to create token: %w", err)
+		}
+	}
+
+	return token.Status.Token, nil
+}
+
+// updateOrCreateSecret creates or update serviceAccount in HCP namespace
```

Comment:
what is this function? we should not need to create or update the serviceaccount within HCP namespace?

Reply:
Just a test and deleted